include/envoy/http/async_client.h

```diff
@@ -205,6 +206,10 @@ class AsyncClient {
       parent_context = v;
       return *this;
     }
+    StreamOptions& setMetadataMatchCriteria(const Router::MetadataMatchCriteria* m) {
```

Comment:
I think this method should follow the example of setHashPolicy and accept a `const envoy::config::core::v3::Metadata&`. That way callers don't need to use the implementation of MetadataMatchCriteria and aren't tempted to provide their own implementation.